File: Notification system/Notification.java
interface NotificationStrategy {
  void sendNotification(String message);
}

class EmailNotification implements NotificationStrategy {
  @override
  void sendNotification(String message) {
    System.out.println("Sending email notification: $message");
  }
}

class SMSNotification implements NotificationStrategy {
  @override
  void sendNotification(String message) {
    System.out.println("Sending SMS notification: $message");
  }
}

class PushNotification implements NotificationStrategy {
  @override
  void sendNotification(String message) {
    System.out.println("Sending push notification: $message");
  }
}

class Notifier {
  private NotificationStrategy notificationStrategy;

  constructor(NotificationStrategy strategy) {
    this.notificationStrategy = strategy;
  }

  notify(string message) {
    if(message == null || message.isEmpty()) {
      throw new IllegalArgumentException("Message cannot be null or empty.");
    }

    notificationStrategy.sendNotification(message);
  }
}

class MultiNotifier {
  private List<NotificationStrategy> notificationStrategies;

  constructor(List<NotificationStrategy> strategies) {
    this.notificationStrategies = strategies;
  }

  notifyAll(string message) {
    if(message == null || message.isEmpty()) {
      throw new IllegalArgumentException("Message cannot be null or empty.");
    }

    for(NotificationStrategy strategy : this.notificationStrategies) {
      strategy.sendNotification(message);
    }
  }
}

Notifier notyfier = new Notifier(new EmailNotification());
notyfier.notify("Hello, this is an email notification!");

MultiNotifier multiNotifier = new MultiNotifier(
  Arrays.asList(new EmailNotification(), 
                new SMSNotification(), 
                new PushNotification())
);

File: Notification system/Notification.ts
interface NotificationStrategy {
  sendNotification(message: string): void;
}

class EmailNotification implements NotificationStrategy {
  sendNotification(message: string): void {
    console.log(`Sending email notification: ${message}`);
  }
}

class SMSNotification implements NotificationStrategy {
  sendNotification(message: string): void {
    console.log(`Sending SMS notification: ${message}`);
  }
}

class PushNotification implements NotificationStrategy {
  sendNotification(message: string): void {
    console.log(`Sending push notification: ${message}`);
  }
}

class Notifier {
  constructor(private notificationStrategy: NotificationStrategy) {}

  notify(message: string): void {
    if(!message) throw new Error("Message cannot be empty");

    this.notificationStrategy.sendNotification(message);
    console.log("Notification sent successfully!");
  }
}

class MultiNotifier {
  constructor(private notificationStrategies: NotificationStrategy[]) {}

  notify(message: string): void {
    if(!message) throw new Error("Message cannot be empty");

    this.notificationStrategies.forEach(
      (strategy) => strategy.sendNotification(message)
    );
    console.log("All notifications sent successfully!");
  }
}

// Usage

const notifier = new Notifier(new SMSNotification());
notifier.notify("You have a new message!");

const emailNotifier = new Notifier(new EmailNotification());
emailNotifier.notify("Your order has been shipped!");

const pushNotifier = new Notifier(new PushNotification());
pushNotifier.notify("You have a new friend request!");